File: engine/process.c
/*
 * process.c - Manage processes within the runtime engine.
 *
 * Copyright (C) 2001  Southern Storm Software, Pty Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "engine.h"

#ifdef	__cplusplus
extern	"C" {
#endif

void ILExecInit(unsigned long maxSize)
{
	/* Initialize the global garbage collector */
	ILGCInit(maxSize);
}

ILExecProcess *ILExecProcessCreate(unsigned long stackSize)
{
	ILExecProcess *process;

	/* Create the process record */
	if((process = (ILExecProcess *)ILGCAllocPersistent
						(sizeof(ILExecProcess))) == 0)
	{
		return 0;
	}

	/* Initialize the fields */
	process->firstThread = 0;
	process->mainThread = 0;
	process->stackSize = ((stackSize < IL_ENGINE_STACK_SIZE)
							? IL_ENGINE_STACK_SIZE : stackSize);
	process->frameStackSize = IL_ENGINE_FRAME_STACK_SIZE;
	process->context = 0;
	process->exitStatus = 0;
	process->coder = 0;
	process->objectClass = 0;
	process->stringClass = 0;
	process->exceptionClass = 0;
	process->clrTypeClass = 0;
	process->outOfMemoryObject = 0;
	ILGetCurrTime(&(process->startTime));
	process->internHash = 0;
	process->reflectionHash = 0;
	process->loadedModules = 0;
	process->entryImage = 0;

	/* Initialize the image loading context */
	if((process->context = ILContextCreate()) == 0)
	{
		ILExecProcessDestroy(process);
		return 0;
	}

	/* Create the "main" thread */
	process->mainThread = ILExecThreadCreate(process);
	if(!(process->mainThread))
	{
		ILExecProcessDestroy(process);
		return 0;
	}

	/* Initialize the CVM coder */
	process->coder = ILCoderCreate(&_ILCVMCoderClass, 100000);
	if(!(process->coder))
	{
		ILExecProcessDestroy(process);
		return 0;
	}

	/* Return the process record to the caller */
	return process;
}

void ILExecProcessDestroy(ILExecProcess *process)
{
	ILLoadedModule *loaded, *nextLoaded;

	/* Destroy the threads associated with the process */
	while(process->firstThread != 0)
	{
		ILExecThreadDestroy(process->firstThread);
	}

	/* Destroy the CVM coder instance */
	ILCoderDestroy(process->coder);

	/* Destroy the image loading context */
	if(process->context)
	{
		ILContextDestroy(process->context);
	}

	/* Destroy the main part of the intern'ed hash table.
	   The rest will be cleaned up by the garbage collector */
	ILGCFreePersistent(process->internHash);

	/* Destroy the main part of the reflection hash table.
	   The rest will be cleaned up by the garbage collector */
	ILGCFreePersistent(process->reflectionHash);

	/* Destroy the loaded module list */
	loaded = process->loadedModules;
	while(loaded != 0)
	{
		if(loaded->handle != 0)
		{
			ILDynLibraryClose(loaded->handle);
		}
		nextLoaded = loaded->next;
		ILFree(loaded);
		loaded = nextLoaded;
	}

	/* Free the process block itself */
	ILGCFreePersistent(process);
}

void ILExecProcessSetLibraryDirs(ILExecProcess *process,
								 char **libraryDirs,
								 int numLibraryDirs)
{
	ILContextSetLibraryDirs(process->context, libraryDirs, numLibraryDirs);
}

ILContext *ILExecProcessGetContext(ILExecProcess *process)
{
	return process->context;
}

ILExecThread *ILExecProcessGetMain(ILExecProcess *process)
{
	return process->mainThread;
}

/*
 * Load standard classes and objects.
 */
static void LoadStandard(ILExecProcess *process, ILImage *image)
{
	ILClass *classInfo;

	if(!(process->outOfMemoryObject))
	{
		/* If this image caused "OutOfMemoryException" to be
		   loaded, then create an object based upon it.  We must
		   allocate this object ahead of time because we won't be
		   able to when the system actually runs out of memory */
		classInfo = ILClassLookupGlobal(ILImageToContext(image),
								        "OutOfMemoryException", "System");
		if(classInfo)
		{
			/* We call the "OutOfMemoryException(int dummy)" constructor
			   in "pnetlib", which sets up the out of memory exception
			   object with the necessary localised string information,
			   but with no stack trace associated with it */
#if 0
			process->outOfMemoryObject =
				ILExecThreadNew(process->mainThread,
								"System.OutOfMemoryException",
								"(Ti)V", (ILVaInt)0);
#endif
			process->outOfMemoryObject =
				_ILEngineAllocObject(process->mainThread, classInfo);
		}
	}

	/* Look for "System.Object" */
	if(!(process->objectClass))
	{
		process->objectClass = ILClassLookupGlobal(ILImageToContext(image),
							        			   "Object", "System");
	}

	/* Look for "System.String" */
	if(!(process->stringClass))
	{
		process->stringClass = ILClassLookupGlobal(ILImageToContext(image),
							        			   "String", "System");
	}

	/* Look for "System.Exception" */
	if(!(process->exceptionClass))
	{
		process->exceptionClass = ILClassLookupGlobal(ILImageToContext(image),
							        			      "Exception", "System");
	}

	/* Look for "System.Reflection.ClrType" */
	if(!(process->clrTypeClass))
	{
		process->clrTypeClass = ILClassLookupGlobal(ILImageToContext(image),
								        "ClrType", "System.Reflection");
	}
}

int ILExecProcessLoadImage(ILExecProcess *process, FILE *file)
{
	ILImage *image;
	int loadError;
	loadError = ILImageLoad(file, 0, process->context, &image,
					   	    IL_LOADFLAG_FORCE_32BIT);
	if(loadError == 0)
	{
		LoadStandard(process, image);
	}
	return loadError;
}

int ILExecProcessLoadFile(ILExecProcess *process, const char *filename)
{
	int error;
	ILImage *image;
	error = ILImageLoadFromFile(filename, process->context, &image,
								IL_LOADFLAG_FORCE_32BIT, 0);
	if(error == 0)
	{
		LoadStandard(process, image);
	}
	return error;
}

int ILExecProcessGetStatus(ILExecProcess *process)
{
	return process->exitStatus;
}

ILMethod *ILExecProcessGetEntry(ILExecProcess *process)
{
	ILImage *image = 0;
	ILToken token;
	while((image = ILContextNextImage(process->context, image)) != 0)
	{
		token = ILImageGetEntryPoint(image);
		if(token && (token & IL_META_TOKEN_MASK) == IL_META_TOKEN_METHOD_DEF)
		{
			process->entryImage = image;
			return ILMethod_FromToken(image, token);
		}
	}
	return 0;
}

int ILExecProcessEntryType(ILMethod *method)
{
	ILType *signature;
	ILType *paramType;
	int entryType;

	/* The method must be static */
	if(!ILMethod_IsStatic(method))
	{
		return IL_ENTRY_INVALID;
	}

	/* The method must have either "void" or "int" as the return type */
	signature = ILMethod_Signature(method);
	if(ILType_HasThis(signature))
	{
		return IL_ENTRY_INVALID;
	}
	paramType = ILTypeGetReturn(signature);
	if(paramType == ILType_Void)
	{
		entryType = IL_ENTRY_NOARGS_VOID;
	}
	else if(paramType == ILType_Int32)
	{
		entryType = IL_ENTRY_NOARGS_INT;
	}
	else
	{
		return IL_ENTRY_INVALID;
	}

	/* The method must either have no args or a single "String[]" arg */
	if(ILTypeNumParams(signature) != 0)
	{
		if(ILTypeNumParams(signature) != 1)
		{
			return IL_ENTRY_INVALID;
		}
		paramType = ILTypeGetParam(signature, 1);
		if(!ILType_IsSimpleArray(paramType))
		{
			return IL_ENTRY_INVALID;
		}
		if(!ILTypeIsStringClass(ILTypeGetElemType(paramType)))
		{
			return IL_ENTRY_INVALID;
		}
		entryType += (IL_ENTRY_ARGS_VOID - IL_ENTRY_NOARGS_VOID);
	}

	/* Return the entry point type to the caller */
	return entryType;
}

long ILExecProcessGetParam(ILExecProcess *process, int type)
{
	switch(type)
	{
		case IL_EXEC_PARAM_GC_SIZE:
		{
			return ILGCGetHeapSize();
		}
		/* Not reached */

		case IL_EXEC_PARAM_MC_SIZE:
		{
			return (long)(ILCoderGetCacheSize(process->coder));
		}
		/* Not reached */
	}
	return -1;
}

#ifdef	__cplusplus
};
#endif

File: System/Diagnostics/Process.cs
/*
 * Process.cs - Implementation of "System.Diagnostics.Process" 
 *
 * Copyright (C) 2002  Southern Storm Software, Pty Ltd.
 * 
 * Contributed by Gopal.V
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

using System;
using System.IO;

namespace System.Diagnostics
{
	[TODO]
	public class Process
	{
		[TODO]
		public Process()
		{
			throw new NotImplementedException(".ctor");
		}

		[TODO]
		public void Close()
		{
			throw new NotImplementedException("Close");
		}

		[TODO]
		public bool CloseMainWindow()
		{
			throw new NotImplementedException("CloseMainWindow");
		}

		[TODO]
		protected virtual void Dispose(bool disposing)
		{
			throw new NotImplementedException("Dispose");
		}

		[TODO]
		public static void EnterDebugMode()
		{
			throw new NotImplementedException("EnterDebugMode");
		}

		[TODO]
		public static Process GetCurrentProcess()
		{
			throw new NotImplementedException("GetCurrentProcess");
		}

		[TODO]
		public static Process GetProcessById(int processId)
		{
			throw new NotImplementedException("GetProcessById");
		}

		[TODO]
		public static Process GetProcessById(int processId, String machineName)
		{
			throw new NotImplementedException("GetProcessById");
		}

		[TODO]
		public static Process[] GetProcesses()
		{
			throw new NotImplementedException("GetProcesses");
		}

		[TODO]
		public static Process[] GetProcesses(String machineName)
		{
			throw new NotImplementedException("GetProcesses");
		}

		[TODO]
		public static Process[] GetProcessesByName(String processName)
		{
			throw new NotImplementedException("GetProcessesByName");
		}

		[TODO]
		public static Process[] GetProcessesByName(String processName, 
													String machineName)
		{
			throw new NotImplementedException("GetProcessesByName");
		}

		[TODO]
		public void Kill()
		{
			throw new NotImplementedException("Kill");
		}

		[TODO]
		public static void LeaveDebugMode()
		{
			throw new NotImplementedException("LeaveDebugMode");
		}

		[TODO]
		protected void OnExited()
		{
			throw new NotImplementedException("OnExited");
		}

		[TODO]
		public void Refresh()
		{
			throw new NotImplementedException("Refresh");
		}

		[TODO]
		public bool Start()
		{
			throw new NotImplementedException("Start");
		}

		[TODO]
		public static Process Start(ProcessStartInfo startInfo)
		{
			throw new NotImplementedException("Start");
		}

		[TODO]
		public static Process Start(String fileName)
		{
			throw new NotImplementedException("Start");
		}

		[TODO]
		public static Process Start(String fileName, String arguments)
		{
			throw new NotImplementedException("Start");
		}

		[TODO]
		public override String ToString()
		{
			throw new NotImplementedException("ToString");
		}

		[TODO]
		public void WaitForExit()
		{
			throw new NotImplementedException("WaitForExit");
		}

		[TODO]
		public bool WaitForExit(int milliseconds)
		{
			throw new NotImplementedException("WaitForExit");
		}

		[TODO]
		public bool WaitForInputIdle()
		{
			throw new NotImplementedException("WaitForInputIdle");
		}

		[TODO]
		public bool WaitForInputIdle(int milliseconds)
		{
			throw new NotImplementedException("WaitForInputIdle");
		}

		[TODO]
		public int BasePriority 
		{ 
			get 
			{
				throw new NotImplementedException("BasePriority");
			}
		}
		[TODO]
		public bool EnableRaisingEvents 
		{ 
			get 
			{
				throw new NotImplementedException("EnableRaisingEvents");
			}

			set 
			{
				throw new NotImplementedException("EnableRaisingEvents");
			}
		}
		[TODO]
		public int ExitCode 
		{ 
			get 
			{
				throw new NotImplementedException("ExitCode");
			}
		}
		[TODO]
		public DateTime ExitTime 
		{ 
			get
			{
				throw new NotImplementedException("ExitTime");
			}
		}
		[TODO]
		public IntPtr Handle 
		{ 
			get
			{
				throw new NotImplementedException("Handle");
			}
		}
		[TODO]
		public int HandleCount 
		{ 
			get
			{
				throw new NotImplementedException("HandleCount");
			}
		}
		[TODO]
		public bool HasExited 
		{ 
			get
			{
				throw new NotImplementedException("HasExited");
			}
		}
		[TODO]
		public int Id 
		{ 
			get
			{
				throw new NotImplementedException("Id");
			}
		}
		[TODO]
		public String MachineName 
		{ 
			get
			{
				throw new NotImplementedException("MachineName");
			}
		}
		[TODO]
		public int MainModule 
		{ 
			get
			{
				throw new NotImplementedException("MainModule");
			}
		}
		[TODO]
		public IntPtr MainWindowHandle 
		{ 
			get
			{
				throw new NotImplementedException("MainWindowHandle");
			}
		}
		[TODO]
		public String MainWindowTitle 
		{ 
			get
			{
				throw new NotImplementedException("MainWindowTitle");
			}
		}
		[TODO]
		public IntPtr MaxWorkingSet 
		{ 
			get
			{
				throw new NotImplementedException("MaxWorkingSet");
			}
		 
			set
			{
				throw new NotImplementedException("MaxWorkingSet");
			}
		}
		[TODO]
		public IntPtr MinWorkingSet 
		{ 
			get
			{
				throw new NotImplementedException("MinWorkingSet");
			}
		 
			set
			{
				throw new NotImplementedException("MinWorkingSet");
			}
		}
		[TODO]
		public int Modules 
		{ 
			get
			{
				throw new NotImplementedException("Modules");
			}
		}
		[TODO]
		public int NonpagedSystemMemorySize 
		{ 
			get
			{
				throw new NotImplementedException("NonpagedSystemMemorySize");
			}
		}
		[TODO]
		public int PagedMemorySize 
		{ 
			get
			{
				throw new NotImplementedException("PagedMemorySize");
			}
		}
		[TODO]
		public int PagedSystemMemorySize 
		{ 
			get
			{
				throw new NotImplementedException("PagedSystemMemorySize");
			}
		}
		[TODO]
		public int PeakPagedMemorySize 
		{ 
			get
			{
				throw new NotImplementedException("PeakPagedMemorySize");
			}
		}
		[TODO]
		public int PeakVirtualMemorySize 
		{ 
			get
			{
				throw new NotImplementedException("PeakVirtualMemorySize");
			}
		}
		[TODO]
		public int PeakWorkingSet 
		{ 
			get
			{
				throw new NotImplementedException("PeakWorkingSet");
			}
		}
		[TODO]
		public bool PriorityBoostEnabled 
		{ 
			get
			{
				throw new NotImplementedException("PriorityBoostEnabled");
			}
		 
			set
			{
				throw new NotImplementedException("PriorityBoostEnabled");
			}
		}
		[TODO]
		public int PriorityClass 
		{ 
			get
			{
				throw new NotImplementedException("PriorityClass");
			}
		 
			set
			{
				throw new NotImplementedException("PriorityClass");
			}
		}
		[TODO]
		public int PrivateMemorySize 
		{ 
			get
			{
				throw new NotImplementedException("PrivateMemorySize");
			}
		}
		[TODO]
		public TimeSpan PrivilegedProcessorTime 
		{ 
			get
			{
				throw new NotImplementedException("PrivilegedProcessorTime");
			}
		}
		[TODO]
		public String ProcessName 
		{ 
			get
			{
				throw new NotImplementedException("ProcessName");
			}
		}
		[TODO]
		public IntPtr ProcessorAffinity 
		{ 
			get
			{
				throw new NotImplementedException("ProcessorAffinity");
			}
		 
			set
			{
				throw new NotImplementedException("ProcessorAffinity");
			}
		}
		[TODO]
		public bool Responding 
		{ 
			get
			{
				throw new NotImplementedException("Responding");
			}
		}
		[TODO]
		public StreamReader StandardError 
		{ 
			get
			{
				throw new NotImplementedException("StandardError");
			}
		}
		[TODO]
		public StreamWriter StandardInput 
		{ 
			get
			{
				throw new NotImplementedException("StandardInput");
			}
		}
		[TODO]
		public StreamReader StandardOutput 
		{ 
			get
			{
				throw new NotImplementedException("StandardOutput");
			}
		}
		[TODO]
		public ProcessStartInfo StartInfo 
		{ 
			get
			{
				throw new NotImplementedException("StartInfo");
			}
		 
			set
			{
				throw new NotImplementedException("StartInfo");
			}
		}
		[TODO]
		public DateTime StartTime 
		{ 
			get
			{
				throw new NotImplementedException("StartTime");
			}
		}
		[TODO]
		public int SynchronizingObject 
		{ 
			get
			{
				throw new NotImplementedException("SynchronizingObject");
			}
		 
			set
			{
				throw new NotImplementedException("SynchronizingObject");
			}
		}
		[TODO]
		public int Threads 
		{ 
			get
			{
				throw new NotImplementedException("Threads");
			}
		}
		[TODO]
		public TimeSpan TotalProcessorTime 
		{ 
			get
			{
				throw new NotImplementedException("TotalProcessorTime");
			}
		}
		[TODO]
		public TimeSpan UserProcessorTime 
		{ 
			get
			{
				throw new NotImplementedException("UserProcessorTime");
			}
		}
		[TODO]
		public int VirtualMemorySize 
		{ 
			get
			{
				throw new NotImplementedException("VirtualMemorySize");
			}
		}
		[TODO]
		public int WorkingSet 
		{ 
			get
			{
				throw new NotImplementedException("WorkingSet");
			}
		}
	}
}//namespace
